### PR TITLE
Add support for d&d multiple widgets

### LINKED ISSUE #3441: Multiple widgets drag&drop with specific selection doesn't work well
## Type of report

Feature request

## Provide detailed reproduction steps (if any)

1. Open `/tests/plugins/widget/manual/clipboardhtml` test.
2. Make selection containing a few widgets, but starting and ending with text.
3. Drag&drop.

### Expected result

Widgets content should change as described in test scenario.

### Actual result

Widgets content doesn't change.

## Other details

* Browser: Firefox
* OS: macOS
* CKEditor version: 4.13
* Installed CKEditor plugins: widget

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
*[#3441](https://github.com/ckeditor/ckeditor4/issues/3441) Added support for dragging and dropping multiple [widgets](https://ckeditor.com/cke4/addon/widget).
```

## What changes did you make?

I've added additional `drop` handler, which fires only if there are more than one widget in the selection or there is something else alongside the widget (e.g. text).

Unfortunately this feature does not work in IE, as it does not support custom MIME types in `DataTransfer`, and in Edge, due to #1850.

Closes #3441.

### CODE REVIEW

**Comment by jacekbogdanski** [comment]
Rebased to the latest upstream.

File: plugins/widget/plugin.js [hunk 1]
```diff
@@ -2613,6 +2613,21 @@
...
 		} );
 
+		// Add support for dropping selection containing more than widget itself
+		// or more than one widget (#3441).
+		editor.on( 'drop', function( evt ) {
+			var dataTransfer = evt.data.dataTransfer,
+				id = dataTransfer.getData( 'cke/widget-id' ),
+				transferType = dataTransfer.getTransferType( editor );
+
+			if ( id !== '' || ( editor.widgets.selected.length > 0 &&
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/widget/plugin.js]: Fixed by merging drop listeners.

File: tests/plugins/widget/manual/clipboardhtmldnd.html [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	<p>many,</p>
+	<p>many,</p>
+	<p>many</p>
+	<p>lines to test scrollhack for IE.</p>
+</div>
+
+<script>
+	if ( CKEDITOR.env.ie ) {
+		bender.ignore();
+	}
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply tests/plugins/widget/manual/clipboardhtmldnd.html]: Actually it was copied from the test from some other d&d tests. I removed the info about scrollhack.

> maybe it's worth to add some comment with issues references, so we can enable this test on IE/Edge once the issues are fixed?

Are they even going to be fixed? Both IE and Edge are dead, so fixing it would require introducing even more hacks to our codebase…

**Comment by Comandeer** [comment]
> Issue when the widget is selected from the bottom

It seems like the issue is reproducible only in Chrome. However I didn't find the source of it yet.

> Issue when the selected widget is inline

This issue is reproducible only if the inline widget is at the end of paragraph being copied. It is probably connected with the fact how we restore selection after pasting. Maybe browsers have problems with putting caret after uneditable element 🤔. It also requires some further investigation.

> Could you verify if a selection is draggable on Safari and if not ignore manual test for this browser?

D&D seems to work fine on Safari.

**Comment by Comandeer** [comment]
> Issue when the selected widget is inline

I've extracted it to #3815, as it's issue inside `editor#insertHtml()` and it's not connected with changes proposed in this PR.

> Issue when the widget is selected from the bottom

This issue is also reproducible without the proposed changes. I've extracted it to #3816.

File: tests/plugins/widget/manual/clipboardhtmldnd.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/widget/manual/clipboardhtmldnd.html]: Probably not, but it gives at least some clue where find the reason of ignoring the test for a full description of the issue. Currently, it's just ignored. Why? Who knows :man_shrugging:

File: tests/plugins/widget/dnd.js [hunk 3]
```diff
@@ -54,11 +54,12 @@
 	var getWidgetById = widgetTestsTools.getWidgetById,
 		assertRelations = lineutilsTestsTools.assertRelations;
 
-	function dragstart( editor, evt, widget ) {
-		var dropTarget = CKEDITOR.plugins.clipboard.getDropTarget( editor );
+	function dragstart( editor, evt, widgetOrNode ) {
```
Comment (jacekbogdanski) [inline tests/plugins/widget/dnd.js]: This change doesn't seem very elegant :thinking: Maybe you could extract this as 2 helpers instead, so whole logic would be encapsulated in one function handling node elements + helper searching for drag handler container inside widget?

The same goes to `dragend` function.

File: tests/plugins/widget/manual/clipboardhtmldnd.html [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	<div data-widget="customwidget">Widget</div>
+	<p>Consectetur dolores voluptatibus illo ipsam eveniet?</p>
+	<div data-widget="customnestedwidget">
+		<p>Widget</p>
+		<div class="ned">
+			<div data-widget="customwidget">Nested widget</div>
+		</div>
+	</div>
+	<p>end here].</p>
+	<p>Some [[inline widget]], too!</p>
```
Comment (jacekbogdanski) [inline tests/plugins/widget/manual/clipboardhtmldnd.html]: Is there any purpose for this inline widget? :thinking: It doesn't count copies which seems invalid during monkey clicking.

File: tests/plugins/widget/manual/clipboardhtmldnd.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+1. Select content inside the editor according to the instructions inside it. Drop the selected content into one of drop zones.
+
+	## Expected
+
+	After dropping widgets' content changes into `Widget copied X times!` where `X` changes after every drop. Note: the same counter is used for all widgets!
+
+	## Unexpected
+
+	Widgets' content doesn't change.
+2. Check if undo is working correctly after performing operations (you can use the button above the editor to reset undo stack)
```
Comment (jacekbogdanski) [inline tests/plugins/widget/manual/clipboardhtmldnd.md]: There is no reset button available.

File: tests/plugins/widget/manual/clipboardhtmldnd.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+@bender-tags: 4.14.0, feature, 3441
+@bender-ui: collapsed
+@bender-ckeditor-plugins: widget, undo, wysiwygarea, toolbar, placeholder
+
+1. Select content inside the editor according to the instructions inside it. Drop the selected content into one of drop zones.
+
+	## Expected
+
+	After dropping widgets' content changes into `Widget copied X times!` where `X` changes after every drop. Note: the same counter is used for all widgets!
```
Comment (jacekbogdanski) [inline tests/plugins/widget/manual/clipboardhtmldnd.md]: It seems a bit broken when the counter is shared between widgets. Maybe we could just replace the counter with some custom message indicating that widgets have been successfully moved?

**Comment by jacekbogdanski** [comment]
Ok, seems like I'm just stupid and didn't find out that you need to wait around 2h pressing mouse (poor one) to trigger D&D on Safari. Thanks @Dumluregn for explanation :smile:

**Comment by Comandeer** [comment]
I've updated unit and manual tests.
